### ISSUE TITLE
Future wish list

### ISSUE THREAD

**Issue by nzakas** [eslint/eslint#5040]
We are almost ready to release 2.0.0, and as such, it's time to start thinking about 3.0.0. As usual, the goal of this issue is to gather feedback on important changes that would be breaking or ambitious. (Other suggestions can just be filed as an issue.)

We are already looking at dropping support for Node < 4 in 3.0.0. 

One thing I'd like to hear is ideas for how can ESLint better support TypeScript and Flow.

**Note:** keep the suggestions as high-level descriptions. This is a brainstorming exercise, we don't want to get into implementation details or plausibility discussions. Let's just see what ideas people have. The best will be explored further. 
##

**Comment by ilyavolodin** [comment]
Well, I think everything that was planned for 2.0 and not implemented will automatically move into 3.0, which means glob configs, selectors, parallel linting, etc.
On top of that, what I would like to have is an investigation into is separating doctrine and jsx parser from the core. It would be great to provide those as parser plugins, so that people who don't care about jsDoc or jsx don't need to spend time waiting for those. It would also allow for other possible formats, like maybe handlebar templates, etc.

**Comment by ilyavolodin** [comment]
I would also like another pass at autofixing. What we have now is great, but very brittle, and only works for a few rules. I think we should also try to expose autofixing to editors (so not autofixing the whole file, but instead doing report first, and then autofixing just the error that was specified), this way editors could have functionality similar to WebStorm/ReSharper, when you not only get notified about errors but you have a nice menu that allows you to fix the issue.

**Comment by ilyavolodin** [comment]
One more item is that I think we need to take another look at local vs. global installation, since it's been confusing users for a while. Would be nice to resolve it in some more user friendly way somehow.

**Comment by platinumazure** [comment]
> On top of that, what I would like to have is an investigation into is separating doctrine and jsx parser from the core. It would be great to provide those as parser plugins, so that people who don't care about jsDoc or jsx don't need to spend time waiting for those. It would also allow for other possible formats, like maybe handlebar templates, etc.

:+1: from me, since that would simplify my microtemplate work.

**Comment by jeffmo** [comment]
I'd love to spend some time to see where Flow can provide any useful bits of its analysis up to ESLint so that plugins can ask for and make use of it. I have some high level ideas around APIs that return maybe-results (depending on presence of a type checker) -- but it would be good to brainstorm a bit.

This would be useful for various lint rules that could make use of types to be more high-fidelity like "Dont use for-in on arrays" or "Don't use strings with setTimeout" or "Call .done() on dangling promises" (etc)

**Comment by johnman** [comment]
It would be good if ESLint 3 supported typings (Facebook flow supports typings and is used by some). My main preference would be typescript support (which has seen increased adoption over the years). This would mean we could drop tslint and use ESLint regardless of whether it was a JS or TypeScript project.

**Comment by nzakas** [comment]
@ilyavolodin 

> On top of that, what I would like to have is an investigation into is separating doctrine and jsx parser from the core. It would be great to provide those as parser plugins, so that people who don't care about jsDoc or jsx don't need to spend time waiting for those. It would also allow for other possible formats, like maybe handlebar templates, etc.

JSX is built into Espree, so separating that out isn't possible. For JSDoc, we could decidecto remove the rule we have now and point people to https://github.com/gajus/eslint-plugin-jsdoc. I think if we roll this up under "decrease download size," we can probably find more packages to remove. 

> I would also like another pass at autofixing. What we have now is great, but very brittle, and only works for a few rules.

Yeah, I definitely want to take another stab at this. 

>  I think we should also try to expose autofixing to editors (so not autofixing the whole file, but instead doing report first, and then autofixing just the error that was specified), this way editors could have functionality similar to WebStorm/ReSharper, when you not only get notified about errors but you have a nice menu that allows you to fix the issue.

I think this is already possible through the API. Of course, we don't handle editor integration directly, but you can get the results with suggested fixes and then do what you want. Is there something in particular you're looking for here?

> One more item is that I think we need to take another look at local vs. global installation, since it's been confusing users for a while. Would be nice to resolve it in some more user friendly way somehow.

Yeah, I'm still not sure if there's a good way to do this (or how necessary it is). But if someone wants to investigate and see if the issues in https://github.com/eslint/eslint/issues/3993 can be addressed, it could be nice.

**Comment by nzakas** [comment]
@jeffmo 

> I'd love to spend some time to see if/where Flow can provide any useful bits of its analysis up to ESLint so that plugins can ask for and make use of it. I have some high level ideas around APIs that return maybe-results (depending on presence of a type checker) -- but it would be good to brainstorm a bit.

Happy to have that discussion.

@johnman You can already use Flow through babel-eslint, and we are working on a TypeScript parser: https://github.com/eslint/typescript-eslint-parser

**Comment by sindresorhus** [comment]
> I'd love to spend some time to see if/where Flow can provide any useful bits of its analysis up to ESLint so that plugins can ask for and make use of it.

:+1: I've had many good rule ideas closed here because of missing type information.

> For JSDoc, we could decidecto remove the rule we have now and point people to https://github.com/gajus/eslint-plugin-jsdoc.

:+1: 

> I would also like another pass at autofixing. What we have now is great, but very brittle, and only works for a few rules.

Relevant: https://github.com/cst/cst

> One more item is that I think we need to take another look at local vs. global installation, since it's been confusing users for a while. Would be nice to resolve it in some more user friendly way somehow.

We've had great success in [AVA](https://github.com/sindresorhus/ava) with using local version if available when executing the global version. As asked for in #3993. It's very simple to implement: https://github.com/sindresorhus/ava/blob/0e6db13a918f0116927299271d91fff590328cb9/cli.js#L6-L14

**Comment by hzoo** [comment]
re: autofixing, When we're finished up with cst in JSCS hopefully we take learnings/figure out how to use it in ESLint?

**Comment by mysticatea** [comment]
@nzakas @ilyavolodin I have a plan to send a PR to add autofixing such as ReSharper to VSCode after [this PR](https://github.com/Microsoft/vscode-eslint/pull/19) was addressed.

This is a early prototype.

![autofix-action](https://cloud.githubusercontent.com/assets/1937871/12542022/28b5a12e-c361-11e5-802c-717ba77e8f53.gif)

I wish to implement autofixing for more rules (e.g. `no-var`, `prefer-const`, ...).

**Comment by ilyavolodin** [comment]
@mysticatea Very nice!

**Comment by nzakas** [comment]
@hzoo I don't see us using CST in ESLint, it doesn't mix well with our architecture and it would limit us to auto-fixing only things CST supports. 

@sindresorhus the big issue we had with #3993 was how to help users differentiate between local and global. Did you look into that at all for Ava? Keep in mind, we have plugins and shareable configs that make our use case a bit more complex.

**Comment by lo1tuma** [comment]
Maybe this idea is a bit too ambitious but I want at least mention it.

Make the cli-engine a separate project, which would be a generic framework for any kind of linters. ESLint does IMHO an awesome job when it comes to file traversal, configurations and plugins. Those things are not really related to JavaScript and could also be useful for other linters. For example [`sass-lint`](https://github.com/sasstools/sass-lint) has a lot of features in common with ESLint:

> [Introduce pluggable config (a la ESLint)](https://github.com/sasstools/sass-lint/issues/400)

Another example would be [`markdownlint`](https://github.com/DavidAnson/markdownlint) which currently doesn’t have a CLI at all.

The benefit of this would be to have a larger community around the framework/cli-engine and we would probably get more contributions.

**Comment by BYK** [comment]
@lo1tuma - I think that's a great idea but why limit with linters? I need something similar for a project of mine and was thinking yanking some code from ESLint :)

**Comment by nzakas** [comment]
@lo1tuma I've actually thought along those lines as well, trying to extract as much from ESLint, making it generic, and publishing it so others could create linters (based on the number of clones we are seeing out there). It would be a massive amount of work, but I think it's feasible.

**Comment by nzakas** [comment]
Here's another crazy thought: what if we completely do away with the configuration cascading in favor of glob-based configuration in files (already proposed #3611)? I think that eliminating the cascading would make this easier, and it would also allow us to move the contents of `.eslintignore` into the configuration file.

**Comment by lo1tuma** [comment]
@nzakas Interesting idea. It would also simplify the [implementation of glob-based configuration](https://github.com/eslint/eslint/commit/4507caaa791a786915c9a72ff93b1349e6df7686).

**Comment by btmills** [comment]
@nzakas :+1: Off the top of my head, a few things moving to glob configuration might improve:

1. Configs can be cached on globs alone instead of some combination of paths + globs, which is where we might end up otherwise, as @lo1tuma pointed out.
2. No longer have to find a clear way to explain "We'll merge all configs up to `/` unless one contains `"root": true` then we'll stop there and only if we don't find anything on the way up to `/` we'll use `~/.eslintrc{.ext}` except remember that if your project is under your home directory and you don't specify `"root": true` then `~/.eslintrc{.ext}` will be taken anyway."
3. I expect it'll reduce the complexity of traversal and config merging because we'll have configs defined before we ever get into traversal.
4. As you pointed out, we might not need `.eslintignore` anymore as long as there's a way to say "don't lint files matching this pattern".

**Comment by ilyavolodin** [comment]
Interesting idea. On one hand, it has potential to significantly simplify our implementation, on the other, it's a pretty dramatic change, which I'm not sure I'm comfortable with without thinking about it too much. I also feel that glob based configs might not be as flexible for minor differences as cascading.

**Comment by nzakas** [comment]
@ilyavolodin don't get too wrapped up on implementation details right now, we're just brainstorming.

**Comment by mysticatea** [comment]
:+1: to eliminating cascading. It would reduce the complexity of traversing and I guess it makes faster.

**Comment by nzakas** [comment]
Another idea: change formatters so they can stream results:
https://github.com/eslint/eslint/issues/4731

**Comment by BYK** [comment]
:+1: for #4731 - sounds quite useful and a worthy upgrade.

**Comment by BenoitZugmeyer** [comment]
I would love to see some work on processors. I already raised the issue in #3422 , and have been told that it would require too much work for now, but it could be a nice addition in the future major version. To further illustrate my needs (and apparently I'm not alone #4153 ), I made a tiny prototype of how it could work : https://github.com/eslint/eslint/compare/master...BenoitZugmeyer:plugin-api-prototype . Essentially:

* have a custom way (= not based on a predefined extension list) to trigger processors;
* expose a `ProcessorContext` in the same spirit as the `RuleContext`, so we can give more informations to processors, and add some APIs in the future (ex: an API to handle rule fixing inside processors).

Those changes may not be requested a lot, but I believe this would attract more people to write better, more complex processors.

**Comment by egoroof** [comment]
I wish async await from ES7 parser support because `babel-eslint` isn't a good solution. Thats all.

**Comment by mysticatea** [comment]
Unfortunately, we have started implementing ES7, but `async/await` dropped from ES7 (probably).
I'm really sad :cry:

**Comment by sindresorhus** [comment]
[Browsers are implementing it now regardless.](https://twitter.com/bterlson/status/692489384654999552) I don't see why ESLint, or rather Acorn, would wait.

**Comment by egoroof** [comment]
oO didn't know about it. What is the cause?

**Comment by ilyavolodin** [comment]
@egoroof Timing. It's still stage 3 proposal, but it needs to be stage 4 by March 1 in order to be accepted. It's very unlikely that it will make it.

@sindresorhus `object.observe` was implemented in Chrome and Opera too. Would not be a good thing to implement something and then have to rip it out.

**Comment by michaelficarra** [comment]
@ilyavolodin It's not unlikely, it's impossible now. All stage advancement has to happen at a face-to-face meeting of TC39. We don't have any more meetings before March 1.

**Comment by sindresorhus** [comment]
@ilyavolodin Highly unlikely that will happen to async functions. Object.observe was a more risky bet. But doesn't matter. We have Babel.

**Comment by hzoo** [comment]
@egoroof babel-eslint would be ok if eslint rules understood async/await syntax. Transforming to es5/es6 is necessary https://github.com/babel/acorn-to-esprima/blob/b20845f56af9a02b4e7ae45a97a53702e480ec34/src/toAST.js#L210-L226 since otherwise other rules would break too. The only current solution is to rewrite rules like in https://github.com/babel/eslint-plugin-babel that can account for async/await in the rules.

**Comment by jeffmo** [comment]
"Dropped from ES7" is probably better worded as "didn't make the train schedule for ES7". Async/await just didn't meet the formal requirements in time -- but this doesn't mean much about its eventual viability..

Async/await will ship. It's just more likely to be ready later this year than in the next month or two. Getting into the spec is a formality, what matters is what browsers and impls do

**Comment by mysticatea** [comment]
Sorry, I'm not used to English. Sure, I believe `async/await` will come to ES8 early time.

**Comment by nzakas** [comment]
Just want to pull back the conversation a bit. This issue is for ESLint features only. What we support language-wise is dependent on what [Espree] (https://github.com/eslint/Espree) supports and so is not tied to ESLint versions.

**Comment by ilyavolodin** [comment]
One more thing that I would want to brainstorm is ability to extend default rules. If we are planning to branch out to support typescript and flow, as well as newer features through babel-eslint, most of our built-in rules would not work correctly out of the box. It would be nice to give plugins ability to somehow extend existing rules. Not sure how to do that right, but I also don't want to see core rules have a huge amount of code to support all possible combinations of features. Would be nice to modularize them somehow.

**Comment by nzakas** [comment]
@ilyavolodin are you talking about just exposing the core rules through an API so people can access them? Or are you talking about more than that?

**Comment by hzoo** [comment]
@ilyavolodin do you mean like in https://github.com/babel/eslint-plugin-babel/pull/44/files - where we have to clone a rule and modify a part of it to take advantage of an experimental feature?

**Comment by ilyavolodin** [comment]
@hzoo Yes, exactly. I think even better example would be react rule that marks `React` as used, so ESLint wouldn't complain about it.
@nzakas I think we talked about this briefly at some point. With es7, typescript and flow, we have a lot of deviations from normal rules ahead of us. I would like to at least think about possibility of making it easier for plugins to either provide an override for a built-in rule (worst case scenario), or ability to somehow extend core rule and provide new functionality on top of existing.
What I want to avoid is having built-in rules have special cases for es7, typescript and flow. It would be a mess.

**Comment by nzakas** [comment]
@ilyavolodin I understand the background, Im asking more about what you think is the solution. In my mind, it's kind of already possible to do what you're saying by grabbing a rule file, so I'm wondering what you think a solution looks like. Also worth mentioning: we would basically never be able to remove rules if we did that, so I'm very interested in how you see this working.

**Comment by ilyavolodin** [comment]
@nzakas I'm not really sure what the solution might be like. Ideally, we would make each rule a "class" with all of the functionality isolated into class methods. That way somebody could inherit from it. And yes, people can currently copy the whole code of the rule and modify it, as @hzoo mentioned, but they can't substitute built-in rule with their own implementation. They can name it the same way, but it would still require plugin prefix.
In general, I'm not sure if there's a good solution to this, I just wanted to throw out a possibility out there and see if somebody has brilliant ideas on how this might work. I just really don't want to have core rules to have a bunch of code forks for typescript and flow...

**Comment by btmills** [comment]
In previous discussions we've tossed around the idea of splitting rules out into separate packages. That approach seems to have worked well for Babel 6. Might that be a small step toward solving the extensibility issue? (Or worth doing even if it doesn't solve extensibility?)

**Comment by ilyavolodin** [comment]
@btmills Oh yeah, I totally forgot about that discussion. Maybe that would help too. Although I have a feeling that people would then expect us to maintain separate plugins for normal js, typescript and flow at the same time.

**Comment by zandaqo** [comment]
How about making autofix more flexible by allowing a user to choose rules that she wants to automatically fix or avoid fixing? Not only would it give more liberty to a user, I believe having this ability in core would allow for more aggressive fixes in rules. We can make the current, "safe", fixes applicably by default, and leave optional the more "dangerous" new fixes for the user to opt-in if she wants.

**Comment by nzakas** [comment]
@zandaqo we can explore letting people indicate which rules to fix. There's a few issues to work through, but seems feasible.

**Comment by nzakas** [comment]
@btmills we can explore that again. I'm not sure how feasible it is without having dedicated maintainers for each package and a way to aggregate rules onto the site for documentation purposes, but if someone wants to investigate, that seems like an interesting direction to go.

**Comment by mathieumg** [comment]
@nzakas Regarding @btmills ' proposal, I imagine the core rules would still all live in this repo but get published individually like Babel core plugins. (Automated using https://github.com/kittens/lerna or something alike)

**Comment by vjeux** [comment]
I'd really love to have a proper integration with github pull requests. I hacked up an integration for react-native but if possible I'd like not to maintain it :p https://github.com/facebook/react-native/blob/master/bots/code-analysis-bot.js

**Comment by vjeux** [comment]
As @mysticatea gif shows well, I think that the next big step is going to integrate with the editor directly. At Facebook, the Nuclide team is thinking a lot about how to make writing JavaScript better and I hope that there can be some common work to be done between the two projects.

**Comment by nzakas** [comment]
@mathieumg that's an interesting concept, but keeping all the rules in this repo doesn't give us the type of flexibility we really want with regards to separate release cycles for rules, separate maintainers, separate issue trackers, etc. We've already had a bit of trouble with `eslint-config-eslint` in this regard, it's unlikely we'll continue down that path.

@vjeux I've seen the bot, but I don't think we have the bandwidth to maintain another project. If you can find a maintainer and want it in the ESLint org, I can definitely arrange that.

**Comment by KevinGrandon** [comment]
It seems there's been a lot of conversation over in https://github.com/eslint/eslint/issues/3458. It would be great if this could make it into 3.0. It would greatly help the extensibility story of eslint.

**Comment by nzakas** [comment]
@KevinGrandon that's definitely on the list of things we're looking at (hence the open issue). Unclear if it would be a breaking change or not at this point.

**Comment by ilyavolodin** [comment]
Few additional suggestions that came to my mind. I've tried doing schema normalization at one point, but we decided to not introduce it back then. I'd like to take another stab at it. Idea is to normalize rule's options based on the schema and provide object that describes rule's setup. I would also like to do a pass-through the rules and set up `defaults` for all schemas, and remove rule's logic for determining defaults (let schema drive that).
I would also like to semi-automate documentation creation based on the predefined template + schema for options.
Metadata propagation to all of the rules is, I think, a given. But we already started that process. Once it's done, I'd like to automate rule list creation and remove hard-coded INDEX.md file.
We also had an open issue at some point to provide command-line option for rule documentation. I think with metadata that might be very powerful, since it could also apply to plugins. We can even add url field into metadata for documentation, and pull it on the fly.
Small internal task would be to investigate AVA for unit testing. Our unit tests take a long time to run currently. It might significantly improve things. Even if we could do it only for rules through RuleTester, that might give us a pretty good boost.

**Comment by SpadeAceman** [comment]
From the comments here, there's some interest in exploring the idea of splitting out rules from core. I propose a single, separate ESLint Rules repo.

Right now, when going through the open issues here for ESLint, it's a mix of:

* bugs & issues with ESLint core, alongside
* a large number of new rule requests and bug reports for existing rules. 

I believe that splitting out rules from the core would provide a better [separation of concerns](https://en.wikipedia.org/wiki/Separation_of_concerns) for everyone involved.

* Combining both here likely adds additional cognitive burden for support, even with tagging.
* The group of people willing & able to submit PRs for ESLint rules is generally going to differ from those submitting PRs for ESLint core (just as those groups will also generally differ from contributors to Espree or Acorn). Yes, there'll be some overlap among maintainers and core contributors, but looking at the larger pool of potential contributors, writing/fixing rules is quite different from working on the core.
* When considering how best to deal with incomplete/problematic issue reports, you can have separate guidelines for rules vs. core, along with (potentially) more aggressive auto-closure of issues in the rules repo.

**Comment by platinumazure** [comment]
I've put this out there before but I think this deserves further consideration: Repository settings/properties, as a way of being able to share basic style properties of your repository and have rules read from them. This simplifies the more complicated auto-fixes for many rules.

Example: `quote-props` auto-fix is impossible to implement because `quote-props` rule does not know whether to use double quotes or single quotes (managed by `quotes` rule).

Solution (possible): Create "repositorySettings" object in root configuration. Set property "quoteStyle" (for example) to "double". `quotes` rule loses its option (this could be done in backwards-compatible way if needed) and refers to repositorySettings.quoteStyle for how to enforce the rule. `quote-props` can then also use repositorySettings.quoteStyle to determine correct auto-fix. In this way, those two rules are "in sync" without actually depending on each other.

Other possible repository settings:
* Line break style
* Space in array literals, array accesses, object literals, parens, etc.
* Spaces around unary and infix operators
* Indentation settings (spaces vs tabs, how much)

Note that most of the example settings above are related to common auto-fix problems we keep running into. Note also that implementing CST-based auto-fix won't really be effective until we can establish a centralized way of determining just how new nodes should be created in a syntax tree and generated as code.

**Comment by hzoo** [comment]
@platinumazure Sounds like a good idea to me and something to explore! Yeah, it sounds like their might be many rules than need the same information - `quotes` and `quote-props` being an example like mentioned. Instead of duplicate data, you take the data from a higher level config option (Could just overwrite the value in the rule if necessary and warn that it is doing so so the user can remove or it's done after a prompt).

**Comment by nevir** [comment]
Support for multiple formatters (and support for file output) would be awesome.  I.e. something like:

```sh
eslint --format stylish --format-to-file junit report/eslint.xml .
```

**Comment by ilyavolodin** [comment]
@nevir That's always been possible. `eslint -f junit -o report/eslint.xml`

**Comment by nevir** [comment]
Well, hell! Thanks for the tip

**Comment by platinumazure** [comment]
@ilyavolodin Not sure we support multiple formatters, do we? (Not that it
would be that hard to use one easily convertible formatter in ESLint and
then use other build steps to copy/transform that into other forms, but
having one formatter for standard out and one or more files would be nice.)

**Comment by nzakas** [comment]
@ilyavolodin I think @nevir means two formatters at the same time, which don't support right now.

**Comment by nevir** [comment]
@nzakas correct

**Comment by mjpearson** [comment]
ESLint CLI's config path argument takes a string file path.  It would be nice if this supported a URL also.

I'm happy to implement this if it gets any votes

**Comment by ilyavolodin** [comment]
What would be a reason to lint remote file through a URL? You could quite easily do it now: `curl http://url/script.js | eslint --stdin` or something like that.

**Comment by mjpearson** [comment]
@ilyavolodin you're completely right, I didn't think of that!  Scratch that suggestion then.

**Comment by platinumazure** [comment]
@ilyavolodin Seems the goal is to have a remote config file, not to lint remote scripts. Correct me if I'm wrong @mjpearson.

**Comment by mjpearson** [comment]
Yes a remote config, for example I want to run an eslint gulp task which fetches the most up to date config from a master repository to run, avoiding different repo's eslint configs from going out of sync.

**Comment by platinumazure** [comment]
Maybe you could accomplish that with a shareable config and have all repos
consume its "latest" tag in devDependencies? You would need to be very
careful about updates to your config to ensure your projects downstream
aren't broken, but it could work. I think it would be better to have NPM
manage this for you rather than ESLint, since NPM has more or less solved
this problem.

**Comment by mjpearson** [comment]
@platinumazure thanks that does make more sense, I'll give it a try.

**Comment by jrencz** [comment]
I added options patching feature suggestion. It may be considered a bullet on this wish list: #5826

**Comment by pedrottimark** [comment]
> Idea is to normalize rule's options based on the schema and provide object that describes rule's setup. I would also like to do a pass-through the rules and set up defaults for all schemas, and remove rule's logic for determining defaults (let schema drive that).

👍 When I studied the declarative schema to find patterns for option descriptions in rule docs, I felt some some mental friction comparing them to the imperative code for default values.

**Comment by platinumazure** [comment]
I want to propose that we address the fact that configurations from directories ancestor to a project root can get pulled in if `root: true` is not specified in the project config.

We've seen a number of issues related to this in chat in the last few days, and I've concluded that requiring `root: true` in the top level config is convenient but surprising. Better would be to require that subdirectory configs which will definitely be extending an ancestor config should say so.

Couple of ways this could be done:

* Require `root: false` in descendant configs (i.e., invert the "default value" of "root")
* Remove `root` support, require descendant configs to extend using a path (`"extends": "../../.eslintrc.json"`)

Note that this problem will also go away if the [single-file glob-based configuration](https://github.com/eslint/eslint/issues/5040#issuecomment-175719032) proposed by @nzakas is accepted.

**Comment by Rantanen** [comment]
(Oops. This ended up being a bit of a longer comment. Didn't want to create a new issue though, given #3993 was already closed. Still wanted to bring this up (again))

About the "global ESLint running local if available" debate (#3993)

Currently plugins are painful. It is far too easy to get into a situation where the global ESLint is completely useless - if the project .eslintrc uses any custom plugins or new rules, the global ESLint won't do anything.

> For JSDoc, we could decide to remove the rule we have now and point people to
> https://github.com/gajus/eslint-plugin-jsdoc. I think if we roll this up under "decrease download
> size," we can probably find more packages to remove.

I'd definitely do something about the local vs global issue before pointing people to use plugins more.

Also if 3.0.0 introduces any breaking changes from 2.x.x for configurations (to either direction), I'd look into resolving this issue with it. Otherwise we'll end up in situations where the user has 3.0.0 and the projects they contribute to are using .eslintrc for 2.x.x - or the other way around.

Looking at #3993, I see three unresolved issues (?): Local vs global plugins, getting users to report the correct version and telling users that they are now running the local version instead of the global one.

### Plugin installation (local vs global)

I'd expect two main scenarios:

#### Local ESLint, local plugins, local .eslintrc

Project specifies their own style options through .eslintrc. Because they have a specific .eslintrc, they want will want a local specific ESLint they know will understand that .eslintrc - users trying to use global ESLint might end up with different warnings.

ie. If the project specifies it's own .eslintrc, there's a chance it uses plugins that it then devDepends on. For this reason alone the local eslint should be used. Even if it doesn't use plugins, the project most likely wants those rules to be used exactly in the way they are implemented in the eslint version the project is using (through a devDependency).

#### Global ESLint, global plugins, global .eslintrc

Project doesn't specify .eslintrc. The user still wants to use ESLint with the project - it might be their own, or then it's a collaborative project that doens't use a linter. (If it uses JSHint or something else, then the user should definitely use that one just to abide by the project rules depites the user preferring ESLint).

If the project doesn't specify .eslintrc, it shouldn't install local version of ESLint and everything should be global.

#### Anything else?

##### Global ESLint, local .eslitnrc/plugins

Doesn't work. Local .eslintrc will refer to the local plugins - but global ESLint will fail (unless the user manually installs these plugins globally).

This is the scenario that is all too common currently: Even with local ESLint available, the `eslint` command most users end up running executes the global ESLint, which then fails as the user hasn't installed the local plugins globally.

##### Local ESLint, global .eslintrc/plugins

If the project defines ESLint as a dependency of some kind, I'd fully expect the project to also define .eslintrc. If it doesn't, there's no guarantee the local ESLint the project defines is compatible with the .eslintrc in the user's home directory.

### Having to chase users for the version information.

As someone pointed out in #3993, there is already a similar problem with `npm run lint` vs `eslint -v` - and there seemed to be several good proposals in #3993 for this: Telling users to run the `eslint -v` in the project directory. This would also take care of the `npm run lint` issue.

### Documenting the local use instead of global

I can appreciate the documentation issue, but I also feel that it shouldn't block the resolution to the whole overall issue of intuitively using the local eslint (especially plugins) when available.

-----------------------------------------------

The more I wrote the summary above, the more I started to think that the location of .eslintrc ties closely to the problem as well. If a project depends on a local ESLint but has no .eslintrc, it makes no real sense to run the local ESLint with the user's default .eslintrc - or if the project has a local .eslintrc, but doesn't depend on ESLint, running the global ESLint with the local .eslintrc is likely to break as well. Especially when that local .eslintrc uses plugins.

ESLint is already using the resolve-ish mechanics for looking up the .eslintrc (check current directory and all parents). I think for this reason the same should be used for finding the ESLint version to run. The other sensible choice would be to always use the user's default .eslintrc when running the global ESLint and only use the project .eslintrc when specifically running the project local ESLint and this just doesn't sound appealing to me.

**Comment by jokeyrhyme** [comment]
Any interest in this? #6651 "proposal: extract CLI and settings handling into separate module(s)"

**Comment by alberto** [comment]
Closing since this is a meta issue
